### PR TITLE
feat: add SCIM provisioning via Okta

### LINKED ISSUE #3501: Automated user provisioning via SCIM
### Problem
Teams who use an identity provider, such as OKTA, will want to manage users directly via that platform. While a user in their IDP org can sign up on Coder with OIDC and be automatically provisioned as a user, they cannot be automatically removed from Coder if they are removed in the IDP. 

### Definition of Done
Admins should be able to remove a user in their IDP (that is connected to Coder via OIDC and SCIM) and have the user automatically be deprovisioned in Coder. This avoids them needing to remove the user in two separate places. 

This should be a paid enterprise feature.

### PR DESCRIPTION
**Pull request by coadler**

### CODE REVIEW

**Review by kylecarbs — COMMENTED** [review]
This doesn't delete users, right? I wonder if people will want/expect that, but it gets a bit tricky because of workspaces...

File: enterprise/coderd/coderd.go [hunk 1]
```diff
@@ -63,6 +63,19 @@ func New(ctx context.Context, options *Options) (*API, error) {
 		})
 	})
 
+	if len(options.ScimAPIKey) != 0 {
+		api.AGPL.RootHandler.Route("/scim/v2", func(r chi.Router) {
+			r.Use(api.scimEnabledMW)
+			r.Post("/Users", api.scimPostUser)
```
[the remark this answers is not in the document]
Comment (coadler) [reply enterprise/coderd/coderd.go:69]: Yeah this is the SCIM spec :(

[the remark this answers is not in the document]
Comment (kylecarbs) [reply enterprise/coderd/coderd.go:69]: craaaazy

**Comment by coadler** [comment]
> This doesn't delete users, right? I wonder if people will want/expect that, but it gets a bit tricky because of workspaces...

Yeah, this is probably fine for now too. People mostly care about deprovisioning access. Deprovisioning resources can definitely be a follow up.

File: codersdk/features.go [hunk 2]
```diff
@@ -17,6 +17,7 @@ const (
 const (
 	FeatureUserLimit = "user_limit"
 	FeatureAuditLog  = "audit_log"
+	FeatureSCIM      = "scim"
 )
 
 var FeatureNames = []string{FeatureUserLimit, FeatureAuditLog}
```
Comment (kylecarbs) [inline codersdk/features.go:23]: This should be added to `FeatureNames`

File: enterprise/cli/server.go [hunk 3]
```diff
@@ -28,6 +30,7 @@ func server() *cobra.Command {
 	})
 	cliflag.BoolVarP(cmd.Flags(), &auditLogging, "audit-logging", "", "CODER_AUDIT_LOGGING", true,
 		"Specifies whether audit logging is enabled.")
+	cliflag.StringVarP(cmd.Flags(), &scimAuthHeader, "scim-auth-header", "", "CODER_SCIM_API_KEY", "", "Enables and sets the authentication header for the built-in SCIM server.")
```
Comment (kylecarbs) [inline enterprise/cli/server.go:33]:
```suggestion
	cliflag.StringVarP(cmd.Flags(), &scimAuthHeader, "scim-auth-header", "", "CODER_SCIM_API_KEY", "", "Enables SCIM and sets the authentication header for the built-in SCIM server.")
```

File: enterprise/coderd/scim.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	})
+}
+
+func (api *API) scimVerifyAuthHeader(r *http.Request) bool {
+	hdr := []byte(r.Header.Get("Authorization"))
+
+	return len(api.SCIMAPIKey) != 0 && subtle.ConstantTimeCompare(hdr, api.SCIMAPIKey) == 1
+}
+
+// getUsers intentionally always returns no users. This is done to always force
```
Comment (kylecarbs) [inline enterprise/coderd/scim.go:42]:
```suggestion
// scimGetUsers intentionally always returns no users. This is done to always force
```

File: enterprise/coderd/coderdenttest/coderdenttest.go [hunk 5]
```diff
@@ -37,6 +37,7 @@ func init() {
 type Options struct {
 	*coderdtest.Options
 	EntitlementsUpdateInterval time.Duration
+	ScimAPIKey                 []byte
```
Comment (kylecarbs) [inline enterprise/coderd/coderdenttest/coderdenttest.go:40]: This should probably be `SCIMAPIKey` for consistency

File: enterprise/coderd/scim.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	_ = handlerutil.WriteError(rw, spec.ErrNotFound)
+}
+
+// We currently use our own struct instead of using the SCIM package. This was
+// done mostly because the SCIM package was almost impossible to use. We only
+// need these fields, so it was much simpler to use our own struct. This was
+// tested only with Okta.
+type ScimUser struct {
```
Comment (kylecarbs) [inline enterprise/coderd/scim.go:79]: `SCIMUser`

File: enterprise/coderd/scim.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+		_ = handlerutil.WriteError(rw, spec.Error{Status: http.StatusBadRequest, Type: "invalidEmail"})
+		return
+	}
+
+	user, _, err := api.AGPL.CreateUser(ctx, api.Database, agpl.CreateUserRequest{
+		CreateUserRequest: codersdk.CreateUserRequest{
+			Username: sUser.UserName,
+			Email:    email,
+		},
+		LoginType: database.LoginTypeOIDC,
```
Comment (kylecarbs) [inline enterprise/coderd/scim.go:133]: Should this be explicitly noted in the flag help? Not sure it's obvious that provisioned users would be OIDC.

**Review by kylecarbs — COMMENTED** [review]
We should add docs in this PR too.

https://coder.com/docs/coder-oss/latest/admin/auth you can add a section for SCIM. I'm sure @bpmct would help!

File: docs/admin/auth.md [hunk 8]
```diff
@@ -74,3 +74,14 @@ CODER_OIDC_CLIENT_SECRET="G0CSP...7qSM"
 Once complete, run `sudo service coder restart` to reboot Coder.
 
 > When a new user is created, the `preferred_username` claim becomes the username. If this claim is empty, the email address will be stripped of the domain, and become the username (e.g. `example@coder.com` becomes `example`).
+
+## SCIM
+
+Coder supports user provisioning and deprovisioning via SCIM 2.0 with header
+authentication. Upon deactivation, users are deactivated within the platform but
+are not deleted. Configure your SCIM application with an auth key and supply it
+to Coder.
```
Comment (bpmct) [inline docs/admin/auth.md:83]:
```suggestion
Coder supports user provisioning and deprovisioning via SCIM 2.0 with header
authentication. Upon deactivation, users are [suspended](userd.md#suspend-a-user)
[suspended](userd.md#suspend-a-user) are not deleted. [Configure](./configure.md)
your SCIM application with an auth key and supply it the Coder server.
```

Comment (bpmct) [reply docs/admin/auth.md:83]: just some wording/consistency changes, thanks for writing the docs!

**Comment by bpmct** [comment]
closes #3501

**Review by kylecarbs — APPROVED** [review]

**Pull request merged**